src/PageContent/Calendar/ContactContent.js

```diff
@@ -1,78 +1,131 @@
 import React, { Component } from 'react';
+import moment from 'moment';
+import { accountsRef, deliveriesRef, donatingAgenciesRef } from '../../FirebaseConfig';
 import { AccountType, StringFormat } from '../../Enums';
 import './Content.css';
 import phone from '../../icons/phone.svg';
-import { formatPhone } from '../../utils/Utils';
+import { formatPhone, objectsAreEqual } from '../../utils/Utils';
 
 class ContactContent extends Component {
     constructor(props) {
         super(props);
         this.state = {
             edit: false,
-            contact: {},
-            memberList: []
+            // 'waiting' is true after 'Saved' is clicked and before changes
+            // from db is propagated down. While it is true, input fields are
+            // disabled
+            waiting: false,
+            savedTimestamp: null,
+            daMemberMap: {},
         };
-        if (this.props.accountType === AccountType.DONATING_AGENCY_MEMBER) {
-            this.state = {
-                contact: {
-                    name: this.props.delivery.donatingAgency.contact.name,
-                    phone: this.props.delivery.donatingAgency.contact.phone,
-                    email: this.props.delivery.donatingAgency.contact.email,
-                    memberList: this.props.delivery.donatingAgency.contact
-                        .memberList
-                }
-            };
-        } else if (this.props.accountType === AccountType.RECEIVING_AGENCY) {
-            this.state = {
-                contact: {
-                    name: this.props.delivery.receivingAgency.contact.name,
-                    phone: this.props.delivery.receivingAgency.contact.phone,
-                    email: this.props.delivery.receivingAgency.contact.email
-                }
-            };
-        }
 
         this.edit = this.edit.bind(this);
-        this.handleChange = this.handleChange.bind(this);
-        this.handleChangeDA = this.handleChangeDA.bind(this);
+        this.saveContact = this.saveContact.bind(this);
+    }
+
+    componentWillReceiveProps(nextProps) {
+        // update state to reflect values properly saved if we were waiting
+        // on the update and the update happened after we wrote to db.
+        if (this.state.waiting && nextProps.delivery.updatedTimestamp > this.state.savedTimestamp) {
```

Comment:
This is because the flow looks like this:

1. `ContactContent` modifies data, calls `db.update()`
2. The update change then triggers the delivery listeners setup in `TileGroup`, and updates the data for that `delivery` in `TileGroup.state.deliveries`
3. The `TileGroup.state` change then gets propagated down as `props` to `ContactContent`

So there is a moment of time between \#2 and \#1 where the `ContactContent.props.delivery.contact` data is still the old version because the change hasn't been propagated down through the props from the listener update yet. Thats why:

- At step 1, when the update finishes, we set `ContactContent.state.savedTimestamp` in `.then()`
- At step 2, when the listener finishes updating the `delivery`, we set `delivery.updatedTimestamp`
- both timestamps are recorded on the same browser, so in terms of time consistency, it should be fine...

So the idea is that we want to make sure the update in `props` that got propagated down happened after we successfully written to db. You can image a case when the user hits `save`, and that turns the `state` into `waiting`, and at the same time another agency updated something on the same `delivery` and that triggered an listener update that got propagated down to `ContactContent`. But we haven't finished writing to the db yet, so the `delivery` in the new `prop` contains the old contact data, and if we didn't do the check, we'd display that old data.

This is still not a great model though, so definitely open to better ideas.